smack-integration-test/src/main/java/org/jivesoftware/smackx/muc/MultiUserChatOccupantIntegrationTest.java

```diff
@@ -101,41 +101,35 @@ public void mucJoinEventOrderingTest() throws Exception {
         messageReflectionSyncPoint.waitForResult(timeout);
 
         final ResultSyncPoint<String, Exception> subjectResultSyncPoint = new ResultSyncPoint<>();
-        List<Object> results = new ArrayList<>();
-
-        mucAsSeenByTwo.addMessageListener(message -> {
-            String body = message.getBody();
-            if (mucMessage.equals(body)) {
-                results.add(body);
-            }
-        });
-
-        mucAsSeenByTwo.addParticipantStatusListener(new ParticipantStatusListener() {
-            @Override public void joined(EntityFullJid participant) {
-                // Ignore self-presence, but record all other participants.
-                final EntityFullJid participantTwo = JidCreate.entityFullFrom(mucAddress, nicknameTwo);
-                if (!participantTwo.equals(participant)) {
-                    results.add(participant);
-                }
+        final List<Stanza> results = new ArrayList<>();
+        conTwo.addStanzaListener(stanza -> {
+            results.add(stanza);
+            if (stanza instanceof Message && ((Message) stanza).getSubject() != null) {
+                subjectResultSyncPoint.signal(((Message) stanza).getSubject());
             }
-        });
-
-        mucAsSeenByTwo.addSubjectUpdatedListener((subject, from) -> {
-            results.add(subject);
-            subjectResultSyncPoint.signal(subject);
-        });
+        }, FromMatchesFilter.create(mucAddress));
 
         try {
-            Presence reflectedJoinPresence = mucAsSeenByTwo.join(nicknameTwo);
-            results.add(reflectedJoinPresence.getFrom()); // Self-presence should be second
+            // Join a MUC room, without invoking the MUC implementation in Smack that would set up 'competing' stanza
```

Comment:
Are you sure that you can't keep using `mucAsSeenByTwo.join(nicknameTwo)` here? I wonder what you mean by 'competing stanza listeners.

The key here seems to be that we now use `XMPPConnectino.addStanzaListener()` and not the MUC listeners anymore. But this should also work if we keep using `mucAsSeenByTwo.join(nicknameTwo)`. At least, I don't see why it should break.

Reply:
The essence of the trouble that we had with this test was apparent parallel execution of stanzas related to MUC when using Smack's MUC API. 

In my refactoring, I didn't distinguish between the MUC listeners and the listeners that are internal to the MUC implementation. Instead, I went with the approach that avoided that completely. That is guaranteed to avoid the issue (now, and in potential future changes). It also adds to the consistency of this particular implementation (in that it uses the low-level API everywhere).